toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,293 @@
+      }
+    },
+    "secret": {
+      "data": {
+        "guid": "4f00b75b-3455-48f1-ba4a-f3b82ebad943"
+      }
+    }
+  },
+  "links": {
+    "self": {
+      "href": "https://api.example.org/v3/service_credential_bindings/dde5ad2a-d8f4-44dc-a56f-0452d744f1c3"
+    },
+    "secret": {
+      "href": "https://api.example.org/v3/secrets/4f00b75b-3455-48f1-ba4a-f3b82ebad943"
+    },
+    "app": {
+      "href": "https://api.example.org/v3/apps/74f7c078-0934-470f-9883-4fddss5b8f13"
+    }
+  }
+}
+```
+
+The `mount_path` field defines the `tmpfs` directory that will contain files for the `secret`'s content. For example, consider an `opaque`-type `secret` with the following `value`:
+
+```json
+"value": {
+  "application.yml" : "---\napplication:\n ...",
+  "secret.txt": "my-secret"
+}
+```
+
+The above `secret` would mount files named `application.yml` and `secret.txt` at the path configured in the binding.
+
+### Creating a `secret`
+
+To create a `secret`, users will use a new CLI command: `cf create-secret`.
+
+```console
+cf create-secret SECRET_NAME -t opaque -p <CREDENTIALS_INLINE>
+```
+
+```console
+cf create-secret SECRET_NAME -t opaque -f <FILE_PATH_TO_CREDENTIALS_FILE>
+```
+
+* `-t` the secret `type`, defaults to `opaque`
+* `-p` reads inline/literal credentials
+* `-f` reads credentials from a file
+
+This command will ultimately create a `POST` request to `/v3/secrets`. Example `curl` invocation:
+
+```console
+curl "https://api.example.org/v3/secrets" \
+  -X POST \
+  -H "Authorization: bearer [token]" \
+  -H "Content-type: application/json" \
+  -d '{
+    "type": "opaque",
+    "name": "my-app-secrets",
+    "value": {
+      "application.yml" : "---\napplication:\n ...",
+      "secret.txt": "my-secret"
+    }
+    "metadata": {
+      "annotations": {},
+      "labels": {}
+    },
+    "relationships": {
+      "space": {
+        "data": {
+          "guid": "7304bc3c-7010-11ea-8840-48bf6bec2d78"
+        }
+      }
+    }
+  }'
+```
+
+### Updating a `secret`
+Users will be able to update a `secret` using similar commands:
+
+```console
+cf update-secret SECRET_NAME SECRET_TYPE -p <CREDENTIALS_INLINE>
+```
+
+```console
+cf update-secret SECRET_NAME SECRET_TYPE -f <FILE_PATH_TO_CREDENTIALS_FILE>
+```
+
+Initially apps will need to be restarted to receive `secret` updates, but eventually we may be able to support updates that don't require a restart.
+The "Update secret" endpoint in the CF API will take a `PATCH` request. Example `curl` invocation:
+
+```console
+curl "https://api.example.org/v3/secrets/[guid]" \
+  -X PATCH \
+  -H "Authorization: bearer [token]" \
+  -H "Content-type: application/json" \
+  -d '{
+    "type": "opaque",
+    "value": {
+      "application.yml" : "---\napplication:\n ...",
+      "secret.txt": "updated-secret-value"
+    }
+    "metadata": {
+      "annotations": {},
+      "labels": {}
+    }
+  }'
+```
+
+### Reading a `secret` and Its Value
```

Comment:
Is there going to be an endpoint to get the full secret "object"?
Is there going to be a CLI command to retrieve the secret?

Reply:
There probably should be a CLI command, but as it stands having a separate endpoint for the non-sensitive bits of the Secret and the Secret value has parity with how env vars and service binding credentials work. This is mostly necessary because of how permissions work in Cloud Controller (we want some roles to be able to see the existence of a Secret / Binding to an app, but not its contents).

The CLI could call both endpoints for `SpaceDeveloper` users and show more details if that's desired.